### ISSUE TITLE
Error: spawn ENAMETOOLONG

### ISSUE THREAD

**Issue by giangde091285** [google/earthengine-api#151]
Dear Earth Engine team's software. I am using the Earth Engine JavaScript API in a Node/Express application to make some calls to generate maps and tiles. And I got the issue like below. Would you please help? Thanks you very much!

`C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15312
            throw callbackError;
            ^

Error: spawn ENAMETOOLONG
    at ChildProcess.spawn (internal/child_process.js:403:11)
    at spawn (child_process.js:553:9)
    at exports.XMLHttpRequest.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\xmlhttprequest\lib\XMLHttpRequest.js:503:22)
    at Object.module$contents$ee$apiclient_apiclient.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15594:13)
    at module$contents$ee$apiclient_EERequestService.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15385:54)
    at module$exports$eeapiclient$promise_api_client.PromiseApiClient.$request (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:7952:69)
    at module$exports$eeapiclient$ee_api_client.ProjectsValueApiClientImpl.compute (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:12385:26)
    at Object.ee.data.computeValue (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:20714:35)
    at ee.Geometry.ee.ComputedObject.getInfo (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:21292:18)
    at Object.<anonymous> (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\ssm_new.js:47:73) {
  errno: -4064,
  code: 'ENAMETOOLONG',
  syscall: 'spawn'
}`

**Comment by donmccurdy** [comment]
Could you share the code that you are running, or a minimal subset of it that reproduces the error?

Note the warnings at the bottom of this page — https://developers.google.com/earth-engine/guides/npm_install#testing-the-installation — methods like `getInfo()` should be avoided in Node.js. Basically, for any method that can be given a callback function, it is important to always use a callback function.

**Comment by giangde091285** [comment]
Thanks @donmccurdy, I attached here the code. And here the command line to excuse the code:
`node ee-runner.js aaa.js`

[aaa.zip](https://github.com/google/earthengine-api/files/5921924/aaa.zip)

**Comment by donmccurdy** [comment]
I'm getting a different error from this code:

```
Error: Image.arrayFlatten: Image must have 1 band, found 3 bands.
    at Object.module$contents$ee$apiclient_apiclient.handleResponse_ (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:15675:9)
    at Object.module$contents$ee$apiclient_apiclient.send (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:15600:49)
    at module$contents$ee$apiclient_EERequestService.send (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:15385:54)
    at module$exports$eeapiclient$promise_api_client.PromiseApiClient.$request (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:7952:69)
    at module$exports$eeapiclient$ee_api_client.ProjectsValueApiClientImpl.compute (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:12385:26)
    at Object.ee.data.computeValue (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:20714:35)
    at ee.Image.ee.ComputedObject.getInfo (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:21292:18)
    at ee.Image.getInfo (/Users/donmccurdy/.nvm/versions/node/v14.14.0/lib/node_modules/ee-runner/node_modules/@google/earthengine/build/main.js:22537:39)
    at Object.<anonymous> (/Users/donmccurdy/Desktop/aaa/aaa.js:132:18)
```

Other than that, the call to `getInfo()` at the end of the script may be part of the problem. If you can use `evaluate(function (info) { ... })` instead, that will often work better in a Node.js environment.

**Comment by giangde091285** [comment]
[aaa_v1.zip](https://github.com/google/earthengine-api/files/5928143/aaa_v1.zip)
Hi @donmccurdy, I did not know what was the reason why you got different error. I made change in the code now, and you might see the same error that I got.
I think the problem happened in Line 115:
`var varX = dir_var.subtract(dir_mean.multiply(dir_mean).multiply(sigmaV)).divide(sigmaV.add(1.0));
`
But I did not know the reason, and I did not know how to debug into this line of code. Could you please rerun with the new version of code and see what will happened. Regarding to the point of using `evaluate(function (info) { ... })`, I did not get your point here, would you please give me an example? Thank a lots!

**Comment by donmccurdy** [comment]
The most recent version of your code seems to be working without errors for me, I get two thumbnail URLs printed, here's one of the results:

![5e34fd0d3f64a1aae162cc81aab60c96-014b8dbea1aed95498017189d079c5b1_getPixels](https://user-images.githubusercontent.com/1848368/106949129-6ef29b00-66e1-11eb-88ba-5c1b66818652.jpg)

Are you still seeing the same error?

***

For example, these changes are likely to work better in a Node.js or Express environment:

```js
// before
var info = image.getInfo();
console.log(info);

// after
image.evaluate(function (info) {
  console.log(info);
});
```

```js
// before
var url = image.visualize().getThumbUrl({...});
console.log(url);

// after
image.visualize().getThumbURL({...}, function (url) {
  console.log(url);
});
```

**Comment by giangde091285** [comment]
It is really strange. I still can see the same error! Do you have any idea why, @donmccurdy? Is there any thing related to node environmental? I am curios which version of node you are using and how did you setup the working environment, I would appreciate if you can support me more! 
```
C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner>node ee-runner.js aaa_v1.js
Running script: aaa_v1.js
C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15312
            throw callbackError;
            ^

Error: spawn ENAMETOOLONG
    at ChildProcess.spawn (internal/child_process.js:403:11)
    at spawn (child_process.js:553:9)
    at exports.XMLHttpRequest.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\xmlhttprequest\lib\XMLHttpRequest.js:503:22)
    at Object.module$contents$ee$apiclient_apiclient.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15594:13)
    at module$contents$ee$apiclient_EERequestService.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15385:54)
    at module$exports$eeapiclient$promise_api_client.PromiseApiClient.$request (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:7952:69)
    at module$exports$eeapiclient$ee_api_client.ProjectsThumbnailsApiClientImpl.create (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:12357:26)
    at Object.ee.data.getThumbId (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:20733:40)
    at ee.Image.getThumbId (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:22572:87)
    at ee.Image.getThumbURL (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:22590:38) {
  errno: -4064,
  code: 'ENAMETOOLONG',
  syscall: 'spawn'
}
```

**Comment by donmccurdy** [comment]
At a technical level, you're probably seeing the error where I am not because the Windows filesystem works differently than the macOS filesystem I'm using. But it's a known issue with the node-xmlhttprequest library (https://github.com/driverdan/node-XMLHttpRequest/issues/176), which is trying to emulate a deprecated browser feature that allows downloading files from remote servers synchronously (i.e., downloading something while freezing the execution of the entire script). Use of that feature is very common in EE scripts, but very uncommon in any other JavaScript environment. That is the reason I am suggesting to use callback functions, if your code is not in the Code Editor.

**Comment by giangde091285** [comment]
Many thanks @donmccurdy. In this case, how can I use callback functions!?
`var varX = dir_var.subtract(dir_mean.multiply(dir_mean).multiply(sigmaV)).divide(sigmaV.add(1.0));
`

**Comment by donmccurdy** [comment]
I don't think that line requires callbacks — are you sure the error traces back to that line?

**Comment by giangde091285** [comment]
Yes, I am very much sure about that line!!!

**Comment by donmccurdy** [comment]
If you comment out lines like this, does the error still occur?

```js
// ...

// "collapse" the stack into a single band image (due to masking, each pixel has just one value in it's directional band, and is otherwise masked)
dir_mean = dir_mean.reduce(ee.Reducer.sum());
dir_var = dir_var.reduce(ee.Reducer.sum());
  
//var image_1 = new ee.Image(dir_var.select(0));
// var url = image_1
//.visualize()
//.getThumbURL({dimensions:'1024x1024', format: 'jpg'});
//console.log(url);



var varX = dir_var.subtract(dir_mean.multiply(dir_mean).multiply(sigmaV)).divide(sigmaV.add(1.0));
//var b = varX.divide(dir_var);
//var kq = dir_mean.add(b.multiply(img.subtract(dir_mean)));
// var result = kq.arrayFlatten([['sum']]);
  
  
// var image = new ee.Image(result.select(0));
// var url = image
// .visualize()
// .getThumbURL({dimensions:'1024x1024', format: 'jpg'});
// console.log(url);
```

I'm not sure how that particular line would cause any network requests, so I think it's more likely that the error would come from the `getThumbURL` calls.

**Comment by giangde091285** [comment]
@donmccurdy  you are right! Here you can see error happened in Line 136
[aaa_v1.zip](https://github.com/google/earthengine-api/files/5928781/aaa_v1.zip)


```
C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner>node ee-runner.js aaa_v1.js
Running script: aaa_v1.js
{
  element: {
    func: { signature_: [Object] },
    args: { input: [Object], bandSelectors: [Object] },
    varName: null
  },
  type: 'EXPORT_IMAGE',
  description: 'test',
  fileFormat: 'GEO_TIFF',
  region: '{"geodesic":false,"type":"Polygon","coordinates":[[[102.73345276165549,10.942290766577212],[105.2987780817077,10.942290766577212],[105.2987780817077,12.894208958090553],[102.73345276165549,12.894208958090553],[102.73345276165549,10.942290766577212]]]}',
  dimensions: '100x100',
  driveFileNamePrefix: 'test'
}
C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15312
            throw callbackError;
            ^

Error: spawn ENAMETOOLONG
    at ChildProcess.spawn (internal/child_process.js:403:11)
    at spawn (child_process.js:553:9)
    at exports.XMLHttpRequest.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\xmlhttprequest\lib\XMLHttpRequest.js:503:22)
    at Object.module$contents$ee$apiclient_apiclient.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15594:13)
    at module$contents$ee$apiclient_EERequestService.send (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:15385:54)
    at module$exports$eeapiclient$promise_api_client.PromiseApiClient.$request (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:7952:69)
    at module$exports$eeapiclient$ee_api_client.ProjectsImageApiClientImpl.export (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:12001:26)
    at Object.ee.data.startProcessing (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\node_modules\@google\earthengine\build\main.js:20951:40)
    at Object.<anonymous> (C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner\aaa_v1.js:136:9)
    at Module._compile (internal/modules/cjs/loader.js:1063:30) {
  errno: -4064,
  code: 'ENAMETOOLONG',
  syscall: 'spawn'
}

C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner>
```

**Comment by donmccurdy** [comment]
My copy of `aaa_v1.js` does not have 136 lines, but `startProcessing(taskId, params, callback)` is another function that requires a callback outside of the Code Editor. It may be easier to write an export this way:

```js
var task = ee.batch.Export.image.toDrive({
 …
});

task.start(() => {
  console.log('Started task #' + task.id);
}, (error) => {
  console.log('Error: ' + error);
});

```

**Comment by giangde091285** [comment]
@donmccurdy there is no error appear when I follow your advice. But nothing is created on my G-driver. What I really want to have is the file locally or in google storage (//gs) bucket.

**Comment by donmccurdy** [comment]
Does this version work? If not, what output and what errors do you see?

```js
console.log('evaluating region...');
dir_var.select(0).geometry().bounds().evaluate(function (region) {
  var taskId = ee.data.newTaskId(1);

  console.log('starting task...');

  ee.data.startProcessing(taskId, {
    element: result.select(0),
    type: 'EXPORT_IMAGE',
    description: 'test',
    fileFormat: 'GEO_TIFF',
    region:region,
    //crs:'EPSG:4326',
    dimensions:'100x100',
    //scale: 30,
    driveFileNamePrefix:'test'
  }, function () {

    console.log('task started!');

  });
});
```

var region = JSON.stringify()
var params =

**Comment by giangde091285** [comment]
Unfortunately not, I just does something likes:

```
C:\Users\dn\AppData\Roaming\npm\node_modules\ee-runner>node ee-runner.js aaa_v1.js
Running script: aaa_v1.js
evaluating region... 
```
Any thought @donmccurdy ?

**Comment by donmccurdy** [comment]
I think this may be an error in ee-runner itself, it is terminating the process without waiting for the results to return. I've filed an issue about this at https://github.com/gee-community/ee-runner/issues/34. You may be able to work around the issue in the meantime by using the installation method documented here: https://developers.google.com/earth-engine/guides/npm_install, rather than going through `ee-runner`. This would require server-side authentication with a service account, explained further on that page.

**Comment by gena** [comment]
This was indeed a but in ee-runner, it was exiting before async calls were finished. Should be fixed in `0.0.36`.

**Comment by donmccurdy** [comment]
@gena thank you! 🙏

@giangde091285 reinstalling ee-runner with something like `npm install ee-runner@0.0.36` should fix the issue mentioned in https://github.com/google/earthengine-api/issues/151#issuecomment-773652042, so that the rest of your code also runs.
